src/core/lib/gprpp/sync.h

```diff
@@ -0,0 +1,126 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* Inline implementation of synchronization primitives, preferred over the C
+   public API for C++ code. */
+
+#ifndef GRPC_CORE_LIB_GPRPP_SYNC_H
+#define GRPC_CORE_LIB_GPRPP_SYNC_H
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/atm.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include <assert.h>
+#include <stdint.h>
+#include <atomic>
+#include <type_traits>
+
+#include "src/core/lib/gprpp/memory.h"
+
+namespace grpc_core {
+
+class RefCount {
```

Comment:
This could use a long comment explaining the intended usage, with some detailed examples.

In particular, I would like to understand why this is necessary, given that we already have a `RefCounted` class:

https://github.com/grpc/grpc/blob/master/src/core/lib/gprpp/ref_counted.h

Reply:
Thank you for the pointer. `RefCount` is a direct replacement for gpr_refcount, with a significantly better performance and binary size.

I removed the `IntrusivelyRefCounted` class because it's a can of worms.